posix: options: add custom Zephyr POSIX subprofile

cfriedt opened the pull request:
* eventfd: bring config to top-level of posix dir, since it is not posix
* posix: separate option groups into c library ext and system interfaces
* posix: c_lib_ext: apply clang-format to getopt sources
* posix: profiles: add custom Zephyr POSIX subprofile
* posix: profiles: make `POSIX_AEP_CHOICE_ZEPHYR` the default
* posix: options: add keep-sorted-start and -stop

Fixes #75843

Should only be merged after
- [x] #88762
- [x] #88767
- [x] #88769
- [x] #90060
- [x] #90096
- [x] #94612
- [x] #95355
- [x] #95224
- [x] #95225
- [x] #95196
- [x] #96589

> [!NOTE]  
> The sources used when `CONFIG_POSIX_C_LANG_SUPPORT_R=y` and `CONFIG_POSIX_C_LIB_EXT=y` can still be 
 overridden with `CONFIG_TC_PROVIDES_POSIX_C_LANG_SUPPORT_R=y` an `CONFIG_TC_PROVIDES_POSIX_C_LIB_EXT=y` (which must be used with `select <option>` in `Kconfig`), as they are primarily meant for toolchain maintainers.

A comparison when building `samples/hello_world` for `native_sim`.

Before:
```shell
wc -l build-main/zephyr/.config
   1295 build-main/zephyr/.config
```

After:
```shell
wc -l build-branch/zephyr/.config
   1220 build-branch/zephyr/.config
```

```shell
diff -Nua build-{main,branch}/zephyr/.config
```
```diff
--- build-main/zephyr/.config   2025-05-24 11:33:53.697011215 -0400
+++ build-branch/zephyr/.config 2025-05-24 11:34:27.140913656 -0400
@@ -900,98 +900,23 @@
 #
 # POSIX API Support
 #
-
-#
-# POSIX Options
-#
 # CONFIG_POSIX_API is not set
-CONFIG_POSIX_AEP_CHOICE_NONE=y
-# CONFIG_POSIX_AEP_CHOICE_BASE is not set
-# CONFIG_POSIX_AEP_CHOICE_PSE51 is not set
-# CONFIG_POSIX_AEP_CHOICE_PSE52 is not set
-# CONFIG_POSIX_AEP_CHOICE_PSE53 is not set
-# CONFIG_POSIX_ASYNCHRONOUS_IO is not set
-# CONFIG_POSIX_BARRIERS is not set
-# CONFIG_POSIX_C_LANG_SUPPORT_R is not set
-# CONFIG_POSIX_C_LIB_EXT is not set
-
-#
-# POSIX device I/O
-#
-# CONFIG_POSIX_DEVICE_IO is not set
-CONFIG_POSIX_OPEN_MAX=4
-# end of POSIX device I/O
-
-# CONFIG_POSIX_FD_MGMT is not set
-# CONFIG_POSIX_FILE_SYSTEM_R is not set
-# CONFIG_POSIX_FILE_SYSTEM is not set
-
-#
-# POSIX memory
-#
-CONFIG_POSIX_PAGE_SIZE=0x40
-# CONFIG_POSIX_SHARED_MEMORY_OBJECTS is not set
-# CONFIG_POSIX_MAPPED_FILES is not set
-# CONFIG_POSIX_MEMORY_PROTECTION is not set
-# end of POSIX memory
-
-# CONFIG_POSIX_MESSAGE_PASSING is not set
-# CONFIG_POSIX_SINGLE_PROCESS is not set
-# CONFIG_POSIX_MULTI_PROCESS is not set
-# CONFIG_POSIX_THREADS is not set
-# CONFIG_POSIX_READER_WRITER_LOCKS is not set
-
-#
-# POSIX scheduler options
-#
-# CONFIG_POSIX_PRIORITY_SCHEDULING is not set
-# end of POSIX scheduler options
-
-# CONFIG_POSIX_SEMAPHORES is not set
-
-#
-# POSIX signals
-#
-# CONFIG_POSIX_REALTIME_SIGNALS is not set
-# CONFIG_POSIX_SIGNALS is not set
-# end of POSIX signals
-
-# CONFIG_POSIX_SPIN_LOCKS is not set
-
-#
-# POSIX synchronized I/O
-#
-# CONFIG_POSIX_FSYNC is not set
-# CONFIG_POSIX_SYNCHRONIZED_IO is not set
-# end of POSIX synchronized I/O
-
-# CONFIG_POSIX_TIMERS is not set
-
-#
-# X/Open system interfaces
-#
-# CONFIG_XSI_REALTIME is not set
-# CONFIG_XSI_SINGLE_PROCESS is not set
-# CONFIG_XOPEN_STREAMS is not set
-# CONFIG_XSI_SYSTEM_LOGGING is not set
-# CONFIG_XSI_THREADS_EXT is not set
-# end of X/Open system interfaces
-
-#
-# Miscellaneous POSIX-related options
-#
-# CONFIG_EVENTFD is not set
-# end of Miscellaneous POSIX-related options
-
+# CONFIG_POSIX_AEP_CHOICE_NONE is not set
+CONFIG_POSIX_AEP_CHOICE_ZEPHYR=y
 CONFIG_TC_PROVIDES_POSIX_C_LANG_SUPPORT_R=y
-# end of POSIX Options
 
 #
-# POSIX Shell Utilities
+# POSIX C Library Extensions
 #
-# end of POSIX Shell Utilities
+CONFIG_POSIX_C_LANG_SUPPORT_R=y
+CONFIG_POSIX_C_LIB_EXT=y
+# CONFIG_GETOPT_LONG is not set
+# end of POSIX C Library Extensions
+
+# CONFIG_POSIX_SYSTEM_INTERFACES is not set
 # end of POSIX API Support
 
+# CONFIG_EVENTFD is not set
 # CONFIG_OPENAMP_RSC_TABLE is not set
 # CONFIG_SMF is not set
 CONFIG_LIBGCC_RTLIB=y
```

Commit:
posix: profiles: add custom Zephyr POSIX subprofile

Add a custom Zephyr POSIX subprofile specifically for enabling the
default features that Zephyr requires as per the coding guidelines.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: separate option groups into c library ext and system interfaces

Separate the POSIX implementation into two categories:

- Extensions to ISO C
- System Interfaces

The first category include standalone functions that generally do not
require OS support or depend on any other features within the POSIX
specification. The Option Groups that comprise this category include

- POSIX_C_LIB_EXT: e.g. strnlen(), fnmatch()
- POSIX_C_LANG_SUPPORT_R: e.g. gmtime_r(), strtok_r()

The second category includes the majority of other POSIX Option Groups
that do require OS support. The latter group may also be categorized
generally as being NATIVE_LIBC_INCOMPATIBLE, although that might
eventually become more granular.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
posix: profiles: make POSIX_AEP_CHOICE_ZEPHYR the default

Default POSIX_AEP_CHOICE to POSIX_AEP_CHOICE_ZEPHYR

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
Probably one or two follow-up PRs can be made to ensure that the `POSIX_C_LIB_EXT` Option Group is complete. I think there are some string manipulation functions missing (not ISO C).

Currently, all of the implementation of `POSIX_C_LANG_SUPPORT_R` is done in `libc/common`, which ... doesn't matter too much, but that could be corrected later if necessary.

Commit:
posix: options: add keep-sorted-start and -stop

Add zephyr-keep-sorted-start and zephyr-keep-sorted-stop
comments.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

Commit:
eventfd: bring config to top-level of posix dir, since it is not posix

The eventfd configuration does not need to be so deeply nested within
POSIX since it does not depend on POSIX completely.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
Shell utilities should probably also be gated by the new Kconfig.

JordanYates commented on lib/posix/Kconfig:
```diff
@@ -1,10 +1,45 @@
+# POSIX System Interfaces
+menuconfig POSIX_SYSTEM_INTERFACES
+    bool "POSIX System Interfaces"
+	depends on !NATIVE_APPLICATION
```
Something wrong with the indentation here

cfriedt replied:
Fixed!

jukkar commented on lib/posix/Kconfig.profile:
```diff
@@ -35,6 +35,28 @@ config POSIX_AEP_CHOICE_NONE
+
+	  These functions are divided into two standalone Option Groups that may be enabled
+	  independently of the remainder of the POSIX API implementation; namely POSIX_C_LIB_EXT and
+	  POSIX_C_LANG_SUPPORT_R. If not referenced by the Zephyr kernel or application, There are no
```
```suggestion
POSIX_C_LANG_SUPPORT_R. If not referenced by the Zephyr kernel or application, there are no
```

cfriedt replied:
Fixed!

jukkar reviewed (commented):
Thanks Chris, this helps to understand what is going on with the POSIX options.

cfriedt commented:
I'll give this a rebase shorty

cfriedt commented:
* Removed commits that should be part of a different PR.

Commit:
posix: c_lib_ext: apply clang-format to getopt sources

The getopt sources were not formatted according to our codying style
conventions, so apply formatting rules as a separate commit.

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
* rebased and split out a few additional PRs
